cli/completion/completion.go

```diff
@@ -0,0 +1,68 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2020 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package completion
+
+import (
+	"bytes"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	completionNoDesc bool //Disable completion description for shells that support it
+)
+
+// NewCommand created a new `version` command
+func NewCommand() *cobra.Command {
+	command := &cobra.Command{
+		Use:       "completion [bash|zsh|fish] [--no-descriptions]",
+		ValidArgs: []string{"bash", "zsh", "fish"},
+		Args:      cobra.ExactArgs(1),
+		Short:     "Generates completion scripts",
+		Long:      "Generates completion scripts for various shells",
+		Example: "  " + os.Args[0] + " completion bash > completion.sh\n" +
+			"  " + "source completion.sh",
+		Run: run,
+	}
+	command.Flags().BoolVar(&completionNoDesc, "no-descriptions", false, "Disable completion description for shells that support it")
+
+	return command
+}
+
+func run(cmd *cobra.Command, args []string) {
+	switch args[0] {
+	case "bash":
+		cmd.Root().GenBashCompletion(os.Stdout)
+		break
+	case "zsh":
+		buf := new(bytes.Buffer)
+		cmd.Root().GenZshCompletion(buf)
+		s := strings.Replace(buf.String(), "#", "", 1)   //remove the comment from compdef
+		r := strings.NewReplacer("[", "\\[", "]", "\\]") //insert escaping before [ and ]
```

Comment:
Here you found a bug in Cobra.  It turns out that `[]` in the description causes problems.  Specifically, it is because of these lines, which use `[]` in their description:
https://github.com/arduino/arduino-cli/blob/6061cbf6dab73f6386c6e6f333a8c145d08315ce/cli/cli.go#L90
https://github.com/arduino/arduino-cli/blob/6061cbf6dab73f6386c6e6f333a8c145d08315ce/cli/cli.go#L92

FYI, there is a PR revamping zsh completion in Cobra, which won't have this problem, but it may take a while to get merged: https://github.com/spf13/cobra/pull/1070

Reply:
I'm following that PR since the beginning with interest! Awesome work and so much to learn. So do you think it's useless to open an issue regarding `[]`?

Reply:
It may help others that hit the problem, so it could prove useful. But it probably won't get fixed since https://github.com/spf13/cobra/pull/1070 should replace the zsh logic.